src/Core/Grid/Query/MerchandiseReturnProductsQueryBuilder.php

```diff
@@ -0,0 +1,202 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Grid\Query;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Class MerchandiseReturnQueryBuilder builds queries for merchandise returns grid data.
+ */
+final class MerchandiseReturnProductsQueryBuilder extends AbstractDoctrineQueryBuilder
+{
+    /**
+     * @var int
+     */
+    private $contextLanguageId;
+
+    /**
+     * @var DoctrineSearchCriteriaApplicatorInterface
+     */
+    private $searchCriteriaApplicator;
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param Connection $connection
+     * @param $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     * @param int $contextLanguageId
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        int $contextLanguageId,
+        /*
+         * @todo
+         * I need RequestStack to get merchandiseReturnId to query builder.
+         * I saw that categoryQueryBuilder uses filters for this purpose and it ends up using
+         * FilterCategorySearchCriteriaListener I wonder should I approach it the same way or stick to request stack?
+         * The filter way could be cleaner, but I still will use Request stack in DefinitionFactory so is it worth to
+         * do it via FilterSearchCriteriaListener here?
```

Comment:
Let's use the `FilterCategorySearchCriteriaListener` strategy again as indeed it's cleaner.
Using the request stack here means "this class cannot be used in CLI context or in webservice context". This is bad. We could use this class to power a `GET /api/order/merchandise_returns` endpoint with filtering capabilities